Backend - Tests - The TestUpgrade/TestVerify test can be flaky

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubeflow_pipelines/4420/kubeflow-pipeline-e2e-test/1299091676467302401
```
integration-test-h6czb-954922127: === RUN   TestUpgrade/TestVerify
integration-test-h6czb-954922127: --- FAIL: TestUpgrade (6.79s)
integration-test-h6czb-954922127:     --- PASS: TestUpgrade/TestPrepare (6.44s)
integration-test-h6czb-954922127:     --- FAIL: TestUpgrade/TestVerify (0.09s)
integration-test-h6czb-954922127:         upgrade_test.go:387: 
integration-test-h6czb-954922127:             	Error Trace:	upgrade_test.go:387
integration-test-h6czb-954922127:             	            				upgrade_test.go:282
integration-test-h6czb-954922127:             	            				upgrade_test.go:64
integration-test-h6czb-954922127:             	Error:      	Not equal: 
integration-test-h6czb-954922127:             	            	expected: &run_model.APIRun{CreatedAt:strfmt.DateTime{wall:0x0, ext:63734161624, loc:(*time.Location)(nil)}, Description:"this is hello world", Error:"", FinishedAt:strfmt.DateTime{wall:0x0, ext:62135596800, loc:(*time.Location)(nil)}, ID:"d809b5a7-4675-4241-bd30-39f61466116f", Metrics:[]*run_model.APIRunMetric(nil), Name:"hello world", PipelineSpec:(*run_model.APIPipelineSpec)(0xc0000e3b00), ResourceReferences:[]*run_model.APIResourceReference{(*run_model.APIResourceReference)(0xc0007c2090), (*run_model.APIResourceReference)(0xc0007c20c0)}, ScheduledAt:strfmt.DateTime{wall:0x0, ext:62135596800, loc:(*time.Location)(nil)}, ServiceAccount:"pipeline-runner", Status:"", StorageState:""}
integration-test-h6czb-954922127:             	            	actual  : &run_model.APIRun{CreatedAt:strfmt.DateTime{wall:0x0, ext:63734161624, loc:(*time.Location)(nil)}, Description:"this is hello world", Error:"", FinishedAt:strfmt.DateTime{wall:0x0, ext:62135596800, loc:(*time.Location)(nil)}, ID:"d809b5a7-4675-4241-bd30-39f61466116f", Metrics:[]*run_model.APIRunMetric(nil), Name:"hello world", PipelineSpec:(*run_model.APIPipelineSpec)(0xc0000e2f00), ResourceReferences:[]*run_model.APIResourceReference{(*run_model.APIResourceReference)(0xc000717500), (*run_model.APIResourceReference)(0xc0007174d0)}, ScheduledAt:strfmt.DateTime{wall:0x0, ext:62135596800, loc:(*time.Location)(nil)}, ServiceAccount:"pipeline-runner", Status:"", StorageState:""}
integration-test-h6czb-954922127:             	            	
integration-test-h6czb-954922127:             	            	Diff:
integration-test-h6czb-954922127:             	            	--- Expected
integration-test-h6czb-954922127:             	            	+++ Actual
integration-test-h6czb-954922127:             	            	@@ -18,3 +18,3 @@
integration-test-h6czb-954922127:             	            	    Key: (*run_model.APIResourceKey)({
integration-test-h6czb-954922127:             	            	-    ID: (string) (len=36) "655fea41-bd80-40e7-a03d-e6911ea763ed",
integration-test-h6czb-954922127:             	            	+    ID: (string) (len=36) "e2f985a9-dc54-4841-833b-b0f5c402f182",
integration-test-h6czb-954922127:             	            	     Type: (run_model.APIResourceType) (len=10) "EXPERIMENT"
integration-test-h6czb-954922127:             	Test:       	TestUpgrade/TestVerify
integration-test-h6czb-954922127: === RUN   TestVisualizationAPI
integration-test-h6czb-954922127: === RUN   TestVisualizationAPI/TestVisualizationAPI
integration-test-h6czb-954922127: --- PASS: TestVisualizationAPI (0.41s)
integration-test-h6czb-954922127:     --- PASS: TestVisualizationAPI/TestVisualizationAPI (0.41s)
integration-test-h6czb-954922127: FAIL
integration-test-h6czb-954922127: FAIL	github.com/kubeflow/pipelines/backend/test/integration	108.971s
```

--- a/backend/test/integration/upgrade_test.go
+++ b/backend/test/integration/upgrade_test.go
@@ -360,6 +360,9 @@ func checkHelloWorldRunDetail(t *testing.T, runDetail *run_model.APIRunDetail) {
 	// Check runtime workflow manifest is not empty
 	assert.Contains(t, runDetail.PipelineRuntime.WorkflowManifest, "whalesay")
 
+	expectedExperimentID := test.GetExperimentIDFromAPIResourceReferences(runDetail.Run.ResourceReferences)
+	require.NotEmpty(t, expectedExperimentID)
+
 	expectedRun := &run_model.APIRun{
 		ID:          runDetail.Run.ID,
 		Name:        "hello world",
@@ -371,7 +374,7 @@ func checkHelloWorldRunDetail(t *testing.T, runDetail *run_model.APIRunDetail) {
 			WorkflowManifest: runDetail.Run.PipelineSpec.WorkflowManifest,
 		},
 		ResourceReferences: []*run_model.APIResourceReference{
-			{Key: &run_model.APIResourceKey{Type: run_model.APIResourceTypeEXPERIMENT, ID: runDetail.Run.ResourceReferences[0].Key.ID},
+			{Key: &run_model.APIResourceKey{Type: run_model.APIResourceTypeEXPERIMENT, ID: expectedExperimentID},
 				Name: "hello world experiment", Relationship: run_model.APIRelationshipOWNER,
 			},
 			{Key: &run_model.APIResourceKey{ID: runDetail.Run.PipelineSpec.PipelineID, Type: run_model.APIResourceTypePIPELINEVERSION},

--- a/backend/test/test_utils.go
+++ b/backend/test/test_utils.go
@@ -28,6 +28,7 @@ import (
 	jobparams "github.com/kubeflow/pipelines/backend/api/go_http_client/job_client/job_service"
 	pipelineparams "github.com/kubeflow/pipelines/backend/api/go_http_client/pipeline_client/pipeline_service"
 	runparams "github.com/kubeflow/pipelines/backend/api/go_http_client/run_client/run_service"
+	"github.com/kubeflow/pipelines/backend/api/go_http_client/run_model"
 	"github.com/kubeflow/pipelines/backend/src/common/client/api_server"
 	"github.com/pkg/errors"
 	"github.com/stretchr/testify/assert"
@@ -95,4 +96,15 @@ func DeleteAllJobs(client *api_server.JobClient, t *testing.T) {
 	for _, j := range jobs {
 		assert.Nil(t, client.Delete(&jobparams.DeleteJobParams{ID: j.ID}))
 	}
+}
+
+func GetExperimentIDFromAPIResourceReferences(resourceRefs []*run_model.APIResourceReference) string {
+	experimentID := ""
+	for _, resourceRef := range resourceRefs {
+		if resourceRef.Key.Type == run_model.APIResourceTypeEXPERIMENT {
+			experimentID = resourceRef.Key.ID
+			break
+		}
+	}
+	return experimentID
 }

